### PR TITLE
Add template class CPLAutoClose

### PR DESCRIPTION
**Pull request by sotex**
## What does this PR do?
Sorry, my English is not good.
In the process of using GDAL, I often encounter places where resources need to be released.

In most cases, std::unique_ptr can be used to automatically release the resources used, but some special places can make it difficult to use or cause the code to be very messy.

For example, when using GDALBuildVRTOptions, I have to do some packaging to use std::unique_ptr, I feel very troublesome, so I simply wrote a template class CPLAutoClose, I think this can be shared.

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS: Windows10 / Linux 4.16.3-1(ArchLinux)
* Compiler: MSVC 14.0 / gcc 7.3.1

### CODE REVIEW

**Comment by rouault** [comment]
@sotex Can you do the little remaining changes I suggested above ?

File: gdal/port/cpl_auto_close.h [hunk 1]
```diff
@@ -49,10 +58,18 @@ class CPLAutoClose {
     CPLAutoClose(const CPLAutoClose&) = delete;
     void operator=(const CPLAutoClose&) = delete;
     public:
+        /**
+         * @brief Constructor.
+         * @param ptr Pointer to the resource object.
+         * @param dr  Resource release(close) function.
```
Comment (rouault) [inline gdal/port/cpl_auto_close.h]: typo: dt

Comment (sotex) [reply gdal/port/cpl_auto_close.h]: You are right, the input is wrong.

Comment (sotex) [reply gdal/port/cpl_auto_close.h]: Already modified, thank you.

**Pull request merged**